File: projects/15-webrtc-sfu/src/sfu.rs
//! The SFU core — **wired** shared state that ties the verticals together, the moral
//! equivalent of project 14's `session.rs`.
//!
//! It owns the room/peer graph and, per peer, the vertical state objects: an [`IceAgent`]
//! (V1), a per-subscriber [`Rewriter`] (V2) + [`LayerSelector`] (V3), and a
//! [`BandwidthEstimator`] (V4). The signaling handlers ([`crate::signaling`]) call the
//! `join_*` / `subscribe` ops to build the graph; the UDP [`pump`](crate::pump) calls the
//! `handle_*` media-plane ops for each datagram. Those media ops do the wired lookup/fan-out
//! and call into the vertical primitives — whose bodies are the `todo!()`s. So this compiles
//! and runs: with no clients it idles; the first STUN check a real browser sends hits the V1
//! `StunMessage::parse` `todo!()` — that panic is your worklist.
//!
//! Locking rule: every method takes the inner `Mutex`, does purely synchronous work, and
//! returns a list of [`Outgoing`] datagrams. The pump performs the actual `send_to` awaits
//! *after* dropping the lock — the lock is never held across `.await`.

use std::collections::HashMap;
use std::net::{IpAddr, SocketAddr};
use std::sync::Mutex;

use bytes::Bytes;
use tracing::{debug, info};

use crate::bwe::BandwidthEstimator;
use crate::error::{Result, SfuError};
use crate::forward::Rewriter;
use crate::ice::{IceAction, IceAgent, StunMessage};
use crate::metrics::{
    BYTES_FORWARDED, ESTIMATED_BITRATE, ICE_NOMINATED, KEYFRAME_REQUESTS, PEERS, ROOMS,
    RTP_DROPPED, RTP_FORWARDED, RTP_RECEIVED, STUN_MESSAGES,
};
use crate::simulcast::{Decision, LayerSelector, SimulcastLayer};
use crate::wire::RtpView;

pub type RoomId = String;
pub type PeerId = u64;

/// A datagram the pump should send on the media socket.
#[derive(Debug)]
pub struct Outgoing {
    pub dst: SocketAddr,
    pub data: Bytes,
}

/// Immutable SFU configuration, shared (behind an `Arc`) by the core + admin server.
#[derive(Debug, Clone)]
pub struct SfuConfig {
    /// Host candidate address advertised to clients in signaling.
    pub public_ip: IpAddr,
    pub media_port: u16,
    pub max_rooms: usize,
    pub max_peers_per_room: usize,
    /// Per-subscriber estimator bounds (bits/sec).
    pub min_bitrate: u32,
    pub start_bitrate: u32,
    pub max_bitrate: u32,
}

impl SfuConfig {
    /// The ICE host candidate the SFU advertises (where clients send their checks + media).
    pub fn media_addr(&self) -> SocketAddr {
        SocketAddr::new(self.public_ip, self.media_port)
    }
}

/// One connected peer and its per-role vertical state.
struct Peer {
    id: PeerId,
    room: RoomId,
    role: Role,
    ice: IceAgent,
    /// The peer's local ICE credentials (the SFU side of the exchange).
    local_ufrag: String,
    /// Publisher: the simulcast layers it announced (empty for a subscriber).
    layers: Vec<SimulcastLayer>,
    /// Subscriber: which publisher it watches, plus its rewriter + selector (None for a publisher).
    subscribes_to: Option<PeerId>,
    rewriter: Option<Rewriter>,
    selector: Option<LayerSelector>,
    bwe: BandwidthEstimator,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    Publisher,
    Subscriber,
}

struct Room {
    peers: Vec<PeerId>,
}

#[derive(Default)]
struct Inner {
    rooms: HashMap<RoomId, Room>,
    peers: HashMap<PeerId, Peer>,
    /// Local ICE ufrag → peer, for routing an inbound STUN check before a pair is nominated.
    by_ufrag: HashMap<String, PeerId>,
    /// Nominated media source address → peer, for routing RTP/RTCP once ICE completes.
    by_addr: HashMap<SocketAddr, PeerId>,
    /// Publisher layer SSRC → publisher peer, for routing an inbound RTP packet to its origin.
    by_ssrc: HashMap<u32, PeerId>,
    next_peer: PeerId,
}

/// The credentials + address handed back to a client from signaling so it can ICE-connect.
#[derive(Debug, serde::Serialize)]
pub struct PeerHandle {
    pub peer_id: PeerId,
    pub ice_ufrag: String,
    pub ice_pwd: String,
    pub media_addr: String,
    /// For a subscriber: the stable SSRC it will receive on.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub out_ssrc: Option<u32>,
}

/// The shared SFU. `Arc<Sfu>` is cloned into the pump task and the signaling router.
pub struct Sfu {
    cfg: SfuConfig,
    inner: Mutex<Inner>,
}

impl Sfu {
    pub fn new(cfg: SfuConfig) -> Self {
        Self {
            cfg,
            inner: Mutex::new(Inner::default()),
        }
    }

    pub fn config(&self) -> &SfuConfig {
        &self.cfg
    }

    // ----------------------------------------------------------------------------------
    // Signaling plane (wired) — build the room/peer graph. Called from HTTP handlers.
    // ----------------------------------------------------------------------------------

    /// Register a **publisher** announcing its simulcast layers; returns its ICE credentials.
    pub fn join_publisher(
        &self,
        room: &str,
        layers: Vec<SimulcastLayer>,
        client_ufrag: String,
    ) -> Result<PeerHandle> {
        let mut inner = self.inner.lock().unwrap();
        self.ensure_room(&mut inner, room)?;
        let (local_ufrag, local_pwd) = gen_credentials();
        let id = inner.next_peer;
        inner.next_peer += 1;

        for layer in &layers {
            inner.by_ssrc.insert(layer.ssrc, id);
        }
        let peer = Peer {
            id,
            room: room.to_string(),
            role: Role::Publisher,
            ice: IceAgent::new(local_ufrag.clone(), local_pwd.clone(), client_ufrag),
            local_ufrag: local_ufrag.clone(),
            layers,
            subscribes_to: None,
            rewriter: None,
            selector: None,
            bwe: BandwidthEstimator::new(
                self.cfg.start_bitrate,
                self.cfg.min_bitrate,
                self.cfg.max_bitrate,
            ),
        };
        self.insert_peer(&mut inner, peer);
        info!(room, peer = id, "publisher joined");
        Ok(PeerHandle {
            peer_id: id,
            ice_ufrag: local_ufrag,
            ice_pwd: local_pwd,
            media_addr: self.cfg.media_addr().to_string(),
            out_ssrc: None,
        })
    }

    /// Register a **subscriber** watching `publisher`; returns its ICE credentials + the
    /// stable SSRC it will receive on. Wires up its rewriter (V2) + layer selector (V3).
    pub fn subscribe(
        &self,
        room: &str,
        publisher: PeerId,
        client_ufrag: String,
    ) -> Result<PeerHandle> {
        let mut inner = self.inner.lock().unwrap();
        let layers = match inner.peers.get(&publisher) {
            Some(p) if p.role == Role::Publisher => p.layers.clone(),
            _ => return Err(SfuError::NotFound(format!("publisher {publisher}"))),
        };
        self.ensure_room(&mut inner, room)?;
        let (local_ufrag, local_pwd) = gen_credentials();
        let out_ssrc = rand::random();
        let id = inner.next_peer;
        inner.next_peer += 1;

        let peer = Peer {
            id,
            room: room.to_string(),
            role: Role::Subscriber,
            ice: IceAgent::new(local_ufrag.clone(), local_pwd.clone(), client_ufrag),
            local_ufrag: local_ufrag.clone(),
            layers: Vec::new(),
            subscribes_to: Some(publisher),
            rewriter: Some(Rewriter::new(out_ssrc)),
            selector: Some(LayerSelector::new(layers)),
            bwe: BandwidthEstimator::new(
                self.cfg.start_bitrate,
                self.cfg.min_bitrate,
                self.cfg.max_bitrate,
            ),
        };
        self.insert_peer(&mut inner, peer);
        info!(room, peer = id, publisher, "subscriber joined");
        Ok(PeerHandle {
            peer_id: id,
            ice_ufrag: local_ufrag,
            ice_pwd: local_pwd,
            media_addr: self.cfg.media_addr().to_string(),
            out_ssrc: Some(out_ssrc),
        })
    }

    /// A JSON-friendly snapshot of the topology for `GET /rooms` and `/status`.
    pub fn topology(&self) -> serde_json::Value {
        let inner = self.inner.lock().unwrap();
        let rooms: Vec<_> = inner
            .rooms
            .iter()
            .map(|(id, r)| {
                serde_json::json!({
                    "room": id,
                    "peers": r.peers.iter().filter_map(|pid| inner.peers.get(pid)).map(|p| {
                        serde_json::json!({ "id": p.id, "role": format!("{:?}", p.role) })
                    }).collect::<Vec<_>>(),
                })
            })
            .collect();
        serde_json::json!({ "rooms": rooms })
    }

    fn ensure_room(&self, inner: &mut Inner, room: &str) -> Result<()> {
        if !inner.rooms.contains_key(room) {
            if inner.rooms.len() >= self.cfg.max_rooms {
                return Err(SfuError::Rejected(format!(
                    "max rooms ({}) reached",
                    self.cfg.max_rooms
                )));
            }
            inner
                .rooms
                .insert(room.to_string(), Room { peers: Vec::new() });
            metrics::gauge!(ROOMS).set(inner.rooms.len() as f64);
        }
        let count = inner.rooms[room].peers.len();
        if count >= self.cfg.max_peers_per_room {
            return Err(SfuError::Rejected(format!(
                "room {room} full ({} peers)",
                self.cfg.max_peers_per_room
            )));
        }
        Ok(())
    }

    fn insert_peer(&self, inner: &mut Inner, peer: Peer) {
        let (id, room, ufrag, role) = (
            peer.id,
            peer.room.clone(),
            peer.local_ufrag.clone(),
            peer.role,
        );
        inner.by_ufrag.insert(ufrag, id);
        inner
            .rooms
            .entry(room)
            .or_insert_with(|| Room { peers: Vec::new() })
            .peers
            .push(id);
        inner.peers.insert(id, peer);
        let role_label = if role == Role::Publisher {
            "publisher"
        } else {
            "subscriber"
        };
        metrics::gauge!(PEERS, "role" => role_label).increment(1.0);
    }

    // ----------------------------------------------------------------------------------
    // Media plane (wired dispatch → vertical primitives). Called from the UDP pump.
    // ----------------------------------------------------------------------------------

    /// Handle an inbound **STUN** datagram (ICE connectivity check). Reaches V1.
    pub fn handle_stun(&self, from: SocketAddr, buf: &[u8]) -> Result<Vec<Outgoing>> {
        // First todo on the media path: parsing the STUN message (V1).
        let msg = StunMessage::parse(buf)?;
        metrics::counter!(STUN_MESSAGES, "kind" => "request").increment(1);

        // Route the check to a peer by the local ufrag in its USERNAME (`<local>:<remote>`).
        let mut inner = self.inner.lock().unwrap();
        let peer_id = username_local_ufrag(&msg)
            .and_then(|u| inner.by_ufrag.get(&u).copied())
            .ok_or_else(|| SfuError::NotFound("no peer for STUN username".into()))?;

        let action = {
            let peer = inner.peers.get_mut(&peer_id).unwrap();
            peer.ice.handle(&msg, from)? // V1
        };
        match action {
            IceAction::Respond(data) => Ok(vec![Outgoing {
                dst: from,
                data: data.into(),
            }]),
            IceAction::Nominated { peer } => {
                inner.by_addr.insert(peer, peer_id);
                metrics::counter!(ICE_NOMINATED).increment(1);
                info!(peer = peer_id, %from, "ICE pair nominated");
                Ok(Vec::new())
            }
            IceAction::Nothing => Ok(Vec::new()),
        }
    }

    /// Handle an inbound **RTP** datagram from a publisher: fan it out to that publisher's
    /// subscribers, each through its selector (V3) + rewriter (V2). Returns the rewritten
    /// datagrams to send.
    pub fn handle_rtp(&self, from: SocketAddr, buf: &[u8]) -> Result<Vec<Outgoing>> {
        let mut inner = self.inner.lock().unwrap();
        metrics::counter!(RTP_RECEIVED).increment(1);

        // The source must be a nominated publisher (ICE done); otherwise drop — an open port
        // takes bytes from anyone.
        let Some(&pub_id) = inner.by_addr.get(&from) else {
            metrics::counter!(RTP_DROPPED, "reason" => "no_route").increment(1);
            return Ok(Vec::new());
        };
        if buf.len() < crate::wire::RTP_MIN_HEADER {
            return Err(SfuError::Truncated {
                need: crate::wire::RTP_MIN_HEADER,
                got: buf.len(),
            });
        }
        let origin_ssrc = u32::from_be_bytes([buf[8], buf[9], buf[10], buf[11]]);
        // Crude keyframe heuristic for the switch boundary (assumes no CSRC/extension); V3
        // refines real H.264/VP8 keyframe detection.
        let is_keyframe = buf
            .get(crate::wire::RTP_MIN_HEADER)
            .map(|b| b & 0x1f == 5)
            .unwrap_or(false);

        // Which subscribers watch this publisher?
        let subs: Vec<PeerId> = inner
            .peers
            .values()
            .filter(|p| p.subscribes_to == Some(pub_id))
            .map(|p| p.id)
            .collect();

        let mut out = Vec::new();
        let mut keyframe_requests: Vec<PeerId> = Vec::new();
        for sub_id in subs {
            let peer = inner.peers.get_mut(&sub_id).unwrap();
            let (selector, rewriter) = match (peer.selector.as_mut(), peer.rewriter.as_mut()) {
                (Some(s), Some(r)) => (s, r),
                _ => continue,
            };
            match selector.on_packet(origin_ssrc, is_keyframe) {
                // V3
                Decision::Forward => {
                    let mut copy = buf.to_vec();
                    if let Some(mut view) = RtpView::new(&mut copy) {
                        rewriter.rewrite(&mut view); // V2
                    }
                    if let Some(dst) = peer.ice.peer() {
                        metrics::counter!(RTP_FORWARDED).increment(1);
                        metrics::counter!(BYTES_FORWARDED).increment(copy.len() as u64);
                        out.push(Outgoing {
                            dst,
                            data: copy.into(),
                        });
                    }
                }
                Decision::Drop => {
                    rewriter.skip(); // V2
                    metrics::counter!(RTP_DROPPED, "reason" => "not_selected").increment(1);
                }
            }
            if selector.wants_keyframe() {
                keyframe_requests.push(pub_id);
            }
        }

        // Ask the publisher for a keyframe if any subscriber's up-switch is waiting on one.
        if !keyframe_requests.is_empty() {
            if let Some(dst) = inner.peers.get(&pub_id).and_then(|p| p.ice.peer()) {
                metrics::counter!(KEYFRAME_REQUESTS).increment(1);
                debug!(publisher = pub_id, "would send PLI/FIR upstream");
                // A real PLI (PSFB, fmt 1) datagram would be built + queued here.
                let _ = dst;
            }
        }
        Ok(out)
    }

    /// Handle an inbound **RTCP** datagram (feedback). Drives the bandwidth estimate (V4) and,
    /// for a NACK, feeds the rewriter's NACK translation (V2). Reaches V4 on a receiver report.
    pub fn handle_rtcp(&self, from: SocketAddr, buf: &[u8]) -> Result<Vec<Outgoing>> {
        if buf.len() < 8 {
            return Err(SfuError::Truncated {
                need: 8,
                got: buf.len(),
            });
        }
        let mut inner = self.inner.lock().unwrap();
        let Some(&peer_id) = inner.by_addr.get(&from) else {
            return Ok(Vec::new());
        };
        let pt = buf[1];
        // Receiver Report (201): the fraction-lost byte sits at offset 12 of the first report
        // block — the loss-based BWE signal (V4). A fuller RTCP compound parser (RR/NACK/TWCC/
        // REMB) is part of the reliability + observability horizontal work.
        if pt == 201 && buf.len() >= 13 {
            let fraction_lost = buf[12] as f64 / 256.0;
            if let Some(peer) = inner.peers.get_mut(&peer_id) {
                let est = peer.bwe.on_loss(fraction_lost); // V4
                metrics::gauge!(ESTIMATED_BITRATE).set(est as f64);
                if let Some(sel) = peer.selector.as_mut() {
                    sel.set_budget(est); // V3 consumes the estimate
                }
            }
        }
        Ok(Vec::new())
    }
}

/// Generate a fresh ICE ufrag + pwd (RFC 5245 wants ≥ 4 / ≥ 22 chars of ICE-allowed text).
fn gen_credentials() -> (String, String) {
    let ufrag = format!("{:08x}", rand::random::<u32>());
    let pwd = format!(
        "{:016x}{:016x}",
        rand::random::<u64>(),
        rand::random::<u64>()
    );
    (ufrag, pwd)
}

/// Extract the local (SFU-side) ufrag from a STUN USERNAME attribute (`<local>:<remote>`).
fn username_local_ufrag(msg: &StunMessage) -> Option<String> {
    msg.attributes.iter().find_map(|a| match a {
        crate::ice::StunAttribute::Username(u) => u.split(':').next().map(|s| s.to_string()),
        _ => None,
    })
}

File: projects/15-webrtc-sfu/src/webrtc_sfu/sfu.py
"""The SFU core — **wired** shared state that ties the verticals together.

It owns the room/peer graph and, per peer, the vertical state objects: an
`IceAgent` (V1), a per-subscriber `Rewriter` (V2) and `LayerSelector` (V3), and
a `BandwidthEstimator` (V4). The signaling handlers (`routes`) call the
`join_publisher` / `subscribe` ops to build the graph; the UDP `pump` calls the
`handle_*` media-plane ops for each datagram. Those media ops do the wired
lookup and fan-out and call into the vertical primitives — whose bodies are the
`NotImplementedError`s.

So this imports, starts and serves. With no clients it idles; the first STUN
check a real browser sends reaches `StunMessage.parse` and raises. That
traceback is your worklist.

## Where the mutex went

The Rust core wrapped its state in a `Mutex`, and the rule was "every method
takes the lock, does purely synchronous work, and returns a list of datagrams;
the pump sends them *after* dropping the lock — never hold it across `.await`".

There is no lock here, and that is not a simplification — it is the same rule
enforced by a different mechanism. These methods contain no `await`, so under
asyncio nothing can interleave with them: a coroutine yields only at an `await`,
and between two of them the event loop is not going anywhere. The mutex existed
to exclude *other threads*, and there are none.

What survives intact is the shape, and it is the part worth keeping: **do the
synchronous work, return the datagrams, let the caller send them.** The moment
one of these methods grows an `await` in the middle, the guarantee evaporates —
a second datagram can be dispatched into a half-updated routing table, and the
bug looks like a peer that intermittently receives someone else's media. If you
ever need I/O in here, return a description of it instead. That is what
`Datagram` is for.

(The other half of the trade: because it is all one thread, a slow method blocks
the whole media plane rather than one lock's worth of it. `handle_rtp` is on the
path for every packet times every subscriber, and it is the first thing to show
up in a `py-spy` profile of the boss fight.)
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass, field
from enum import StrEnum

import structlog
from pydantic import BaseModel

from .bwe import BandwidthEstimator
from .config import Settings
from .errors import NotFoundError, RejectedError, TruncatedError
from .forward import Rewriter
from .ice import IceAgent, StunMessage, Username
from .metrics import (
    BYTES_FORWARDED_TOTAL,
    ESTIMATED_BITRATE_BPS,
    ICE_NOMINATED_TOTAL,
    KEYFRAME_REQUESTS_TOTAL,
    PEERS,
    ROOMS,
    RTP_DROPPED_TOTAL,
    RTP_FORWARDED_TOTAL,
    RTP_RECEIVED_TOTAL,
    STUN_MESSAGES_TOTAL,
)
from .simulcast import Decision, LayerSelector, SimulcastLayer
from .wire import RTP_MIN_HEADER, Address, RtpPacket

__all__ = ["Datagram", "PeerHandle", "PeerView", "Role", "RoomView", "Sfu", "Topology"]

logger = structlog.get_logger(__name__)

RTCP_RECEIVER_REPORT = 201
"""RTCP packet type for a Receiver Report — the loss-based BWE signal (V4)."""


class Role(StrEnum):
    PUBLISHER = "publisher"
    SUBSCRIBER = "subscriber"


@dataclass(frozen=True, slots=True)
class Datagram:
    """A datagram the pump should send on the media socket."""

    payload: bytes
    dst: Address


class PeerHandle(BaseModel):
    """The credentials + address signaling hands back so a client can ICE-connect."""

    peer_id: int
    ice_ufrag: str
    ice_pwd: str
    media_addr: str
    out_ssrc: int | None = None
    """For a subscriber: the stable SSRC it will receive on."""


class PeerView(BaseModel):
    """One peer in the topology snapshot."""

    id: int
    role: Role


class RoomView(BaseModel):
    room: str
    peers: list[PeerView]


class Topology(BaseModel):
    """What `GET /rooms` and `GET /status` publish about the live graph.

    Deliberately thin: peer ids and roles, no ICE credentials, no nominated
    addresses, no SSRCs. This endpoint is unauthenticated, and a topology view
    that leaks a `pwd` hands anyone who reads it the ability to complete ICE as
    that peer.
    """

    rooms: list[RoomView]


@dataclass(slots=True)
class Peer:
    """One connected peer and its per-role vertical state."""

    id: int
    room: str
    role: Role
    ice: IceAgent
    local_ufrag: str
    bwe: BandwidthEstimator

    layers: list[SimulcastLayer] = field(default_factory=list[SimulcastLayer])
    """Publisher: the simulcast layers it announced. Empty for a subscriber."""

    subscribes_to: int | None = None
    """Subscriber: the publisher it watches. `None` for a publisher."""

    rewriter: Rewriter | None = None
    selector: LayerSelector | None = None


class Sfu:
    """The shared SFU. One instance per process, held on the app state."""

    __slots__ = (
        "_by_addr",
        "_by_ssrc",
        "_by_ufrag",
        "_next_peer",
        "_peers",
        "_rooms",
        "_subscribers_of",
        "config",
    )

    def __init__(self, config: Settings) -> None:
        self.config = config
        self._rooms: dict[str, list[int]] = {}
        self._peers: dict[int, Peer] = {}
        self._by_ufrag: dict[str, int] = {}
        """Local ICE ufrag -> peer, to route an inbound check before nomination."""
        self._by_addr: dict[Address, int] = {}
        """Nominated media source address -> peer, to route RTP/RTCP after ICE."""
        self._by_ssrc: dict[int, int] = {}
        """Publisher layer SSRC -> publisher, to find an inbound packet's origin."""
        self._subscribers_of: dict[int, list[int]] = {}
        """Publisher -> its subscribers.

        Maintained rather than derived, and that is a CPython decision. The Rust
        scanned every peer per packet to find the subscribers of one publisher;
        at 1500 packets/sec against a full room that is a linear scan a hundred
        thousand times a second, and an interpreted linear scan costs what a
        compiled one does not. The fan-out itself is unavoidably
        O(subscribers) — the *lookup* does not have to be.
        """
        self._next_peer = 1

    # ------------------------------------------------------------------ #
    # Signaling plane (wired) — build the room/peer graph.
    # ------------------------------------------------------------------ #

    def join_publisher(
        self,
        room: str,
        layers: list[SimulcastLayer],
        client_ufrag: str,
    ) -> PeerHandle:
        """Register a publisher announcing its simulcast layers."""
        self._ensure_room(room)
        local_ufrag, local_pwd = _generate_credentials()
        peer_id = self._next_peer
        self._next_peer += 1

        peer = Peer(
            id=peer_id,
            room=room,
            role=Role.PUBLISHER,
            ice=IceAgent(local_ufrag, local_pwd, client_ufrag),
            local_ufrag=local_ufrag,
            bwe=self._new_estimator(),
            layers=layers,
        )
        for layer in layers:
            self._by_ssrc[layer.ssrc] = peer_id
        self._subscribers_of.setdefault(peer_id, [])
        self._insert_peer(peer)

        logger.info("publisher joined", room=room, peer=peer_id, layers=len(layers))
        return PeerHandle(
            peer_id=peer_id,
            ice_ufrag=local_ufrag,
            ice_pwd=local_pwd,
            media_addr=_format_addr(self.config.media_addr),
        )

    def subscribe(self, room: str, publisher: int, client_ufrag: str) -> PeerHandle:
        """Register a subscriber watching `publisher`.

        Wires up its rewriter (V2) and layer selector (V3), and returns the
        stable SSRC it will receive on for the whole session — including across
        every layer switch, which is the point.
        """
        origin = self._peers.get(publisher)
        if origin is None or origin.role is not Role.PUBLISHER:
            raise NotFoundError(f"no publisher {publisher}")

        self._ensure_room(room)
        local_ufrag, local_pwd = _generate_credentials()
        out_ssrc = secrets.randbits(32)
        peer_id = self._next_peer
        self._next_peer += 1

        peer = Peer(
            id=peer_id,
            room=room,
            role=Role.SUBSCRIBER,
            ice=IceAgent(local_ufrag, local_pwd, client_ufrag),
            local_ufrag=local_ufrag,
            bwe=self._new_estimator(),
            subscribes_to=publisher,
            rewriter=Rewriter(out_ssrc),
            selector=LayerSelector(origin.layers),
        )
        self._insert_peer(peer)
        self._subscribers_of.setdefault(publisher, []).append(peer_id)

        logger.info("subscriber joined", room=room, peer=peer_id, publisher=publisher)
        return PeerHandle(
            peer_id=peer_id,
            ice_ufrag=local_ufrag,
            ice_pwd=local_pwd,
            media_addr=_format_addr(self.config.media_addr),
            out_ssrc=out_ssrc,
        )

    def topology(self) -> Topology:
        """A snapshot of the live graph for `GET /rooms` and `GET /status`."""
        return Topology(
            rooms=[
                RoomView(
                    room=room,
                    peers=[
                        PeerView(id=peer.id, role=peer.role)
                        for peer_id in peer_ids
                        if (peer := self._peers.get(peer_id)) is not None
                    ],
                )
                for room, peer_ids in self._rooms.items()
            ]
        )

    def _new_estimator(self) -> BandwidthEstimator:
        return BandwidthEstimator(
            self.config.start_bitrate,
            self.config.min_bitrate,
            self.config.max_bitrate,
        )

    def _ensure_room(self, room: str) -> None:
        """Create the room if new, and enforce both caps.

        Both limits are the "bounded everything" checklist item: an open
        signaling API means room names and join counts are attacker-chosen, and
        a room table with no cap is a dictionary an anonymous caller can grow
        until the process dies.
        """
        if room not in self._rooms:
            if len(self._rooms) >= self.config.max_rooms:
                raise RejectedError(f"max rooms ({self.config.max_rooms}) reached")
            self._rooms[room] = []
            ROOMS.set(len(self._rooms))
        if len(self._rooms[room]) >= self.config.max_peers_per_room:
            raise RejectedError(f"room full ({self.config.max_peers_per_room} peers)")

    def _insert_peer(self, peer: Peer) -> None:
        self._by_ufrag[peer.local_ufrag] = peer.id
        self._rooms.setdefault(peer.room, []).append(peer.id)
        self._peers[peer.id] = peer
        PEERS.labels(role=peer.role.value).inc()

    # ------------------------------------------------------------------ #
    # Media plane (wired dispatch -> vertical primitives).
    # ------------------------------------------------------------------ #

    def handle_stun(self, source: Address, datagram: bytes) -> list[Datagram]:
        """Handle an inbound STUN datagram (an ICE connectivity check). Reaches V1."""
        # The first todo on the media path: parsing the message (V1).
        message = StunMessage.parse(datagram)
        STUN_MESSAGES_TOTAL.labels(kind="request").inc()

        # Route the check to a peer by the local ufrag in its USERNAME.
        ufrag = _username_local_ufrag(message)
        peer_id = self._by_ufrag.get(ufrag) if ufrag is not None else None
        if peer_id is None:
            raise NotFoundError("no peer for STUN username")

        result = self._peers[peer_id].ice.handle(message, source)  # V1
        if result.nominated is not None:
            self._by_addr[result.nominated] = peer_id
            ICE_NOMINATED_TOTAL.inc()
            logger.info("ICE pair nominated", peer=peer_id, source=_format_addr(result.nominated))
        if result.response is None:
            return []
        return [Datagram(payload=result.response, dst=source)]

    def handle_rtp(self, source: Address, datagram: bytes) -> list[Datagram]:
        """Fan an inbound RTP packet out to the publisher's subscribers.

        Each subscriber gets it through its own selector (V3) and rewriter (V2).
        Returns the rewritten datagrams for the pump to send.
        """
        RTP_RECEIVED_TOTAL.inc()

        # The source must have completed ICE; an open port takes bytes from
        # anyone, and forwarding for an unauthenticated sender would make this
        # SFU a reflector.
        publisher_id = self._by_addr.get(source)
        if publisher_id is None:
            RTP_DROPPED_TOTAL.labels(reason="no_route").inc()
            return []
        if len(datagram) < RTP_MIN_HEADER:
            raise TruncatedError(need=RTP_MIN_HEADER, got=len(datagram))

        origin_ssrc = int.from_bytes(datagram[8:12], "big")
        # A crude keyframe heuristic for the switch boundary: H.264 NAL type 5,
        # assuming no CSRCs and no header extension. Real per-codec keyframe
        # detection (H.264 IDR vs VP8's `P` bit) is part of V3.
        is_keyframe = len(datagram) > RTP_MIN_HEADER and (datagram[RTP_MIN_HEADER] & 0x1F) == 5

        outgoing: list[Datagram] = []
        keyframe_owed = False
        for sub_id in self._subscribers_of.get(publisher_id, ()):
            peer = self._peers.get(sub_id)
            if peer is None or peer.selector is None or peer.rewriter is None:
                continue

            if peer.selector.on_packet(origin_ssrc, is_keyframe) is Decision.FORWARD:  # V3
                destination = peer.ice.peer
                if destination is not None:
                    # One mutable copy per subscriber: each gets its own
                    # sequence number, so they cannot share a buffer. This is
                    # the allocation the boss fight's CPU budget is spent on —
                    # see `wire.py`.
                    buf = bytearray(datagram)
                    peer.rewriter.rewrite(RtpPacket(buf))  # V2
                    RTP_FORWARDED_TOTAL.inc()
                    BYTES_FORWARDED_TOTAL.inc(len(buf))
                    outgoing.append(Datagram(payload=bytes(buf), dst=destination))
            else:
                peer.rewriter.skip()  # V2 — keeps the outbound line gapless
                RTP_DROPPED_TOTAL.labels(reason="not_selected").inc()

            keyframe_owed = keyframe_owed or peer.selector.wants_keyframe

        if keyframe_owed:
            publisher = self._peers.get(publisher_id)
            if publisher is not None and publisher.ice.peer is not None:
                KEYFRAME_REQUESTS_TOTAL.inc()
                logger.debug("would send PLI/FIR upstream", publisher=publisher_id)
                # TODO(protocols): build a real PLI (RTCP PSFB, fmt 1) here and
                # append it to `outgoing`. Until then an up-switch never gets
                # the keyframe it is waiting for, so V3 commits only against a
                # publisher that sends them on its own schedule.
        return outgoing

    def handle_rtcp(self, source: Address, datagram: bytes) -> list[Datagram]:
        """Handle inbound RTCP feedback. Reaches V4 on a receiver report."""
        if len(datagram) < 8:
            raise TruncatedError(need=8, got=len(datagram))
        peer_id = self._by_addr.get(source)
        if peer_id is None:
            return []

        # Receiver Report: the fraction-lost byte sits at offset 12, in the
        # first report block. A full compound parser (RR / NACK / TWCC / REMB)
        # is the reliability + observability horizontal work; this one byte is
        # enough to close the loop into the estimator.
        if datagram[1] == RTCP_RECEIVER_REPORT and len(datagram) >= 13:
            peer = self._peers.get(peer_id)
            if peer is not None:
                estimate = peer.bwe.on_loss(datagram[12] / 256.0)  # V4
                ESTIMATED_BITRATE_BPS.set(estimate)
                if peer.selector is not None:
                    peer.selector.set_budget(estimate)  # V3 consumes the estimate
        return []


def _generate_credentials() -> tuple[str, str]:
    """A fresh ICE ufrag + pwd (RFC 5245 wants >= 4 and >= 22 chars of ICE-safe text).

    `secrets`, not `random`. This is the one place in the project where the
    module choice is a security bug rather than a style preference: Python's
    `random` is a Mersenne Twister, and 624 observed outputs reveal its entire
    internal state — and an SFU hands out a ufrag to every caller who asks. With
    `random`, predicting the next peer's `pwd` is a solved exercise, and that
    `pwd` is the *only* thing standing between an attacker and a nominated ICE
    pair. Hex keeps every character inside the ICE-allowed set.
    """
    return secrets.token_hex(4), secrets.token_hex(16)


def _format_addr(address: Address) -> str:
    """Render an address for signaling and logs, bracketing IPv6 literals."""
    host, port = address
    return f"[{host}]:{port}" if ":" in host else f"{host}:{port}"


def _username_local_ufrag(message: StunMessage) -> str | None:
    """Pull the local (SFU-side) ufrag out of a STUN USERNAME (`<local>:<remote>`)."""
    for attribute in message.attributes:
        if isinstance(attribute, Username):
            return attribute.value.split(":", 1)[0]
    return None
